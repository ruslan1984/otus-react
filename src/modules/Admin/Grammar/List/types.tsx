export type GrammarList = Array<GrammarListItem>;
export type GrammarListItem = {
  id: number;
  name: string;
};
export type GrammarListType = {
  list: GrammarList;
  loading: boolean;
};
