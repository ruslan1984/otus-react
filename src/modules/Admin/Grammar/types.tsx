export type grammar = {
  id: number;
  name: string;
  text: string;
};
export type grammarList = {
  id: number;
  name: string;
};
