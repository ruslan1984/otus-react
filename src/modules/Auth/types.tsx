export enum CheckState {
  initiated,
  succeed,
  failed,
  loading,
}
export type Login = {
  user: string;
  password: string;
};
export type Auth = {
  user: string;
  password: string;
  status: CheckState;
};
