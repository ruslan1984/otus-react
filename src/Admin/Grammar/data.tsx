let data = [
    {
        id: 1,
        name: "Страница 1",
        text: "Контент 1",
    },
    {
        id: 2,
        name: "Страница 2",
        text: "Контент 2",
    },
];

export function grammarList() {
    const result = data.map((item) => {
        return {
            id: item.id,
            name: item.name,
        };
    });

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(result);
        }, 1000);
    });
}

export const grammarDetail = (id: number) => {
    let result: any;
    data.forEach((item) => {
        if (id == item.id) {
            result = {
                name: item.name,
                text: item.text,
            };
        }
    });

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(result);
        }, 1000);
    });
};
export function updateDetail(id: number, update_data: any) {
    data = data.map((item) => {
        if (id == item.id) {
            return {
                id: item.id,
                name: update_data.name,
                text: update_data.text,
            };
        }
        return item;
    });
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve(true);
        }, 1000);
    });
}
