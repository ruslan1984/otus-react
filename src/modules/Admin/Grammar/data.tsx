import * as R from "ramda";

import { grammar } from "./types";
import { sleep } from "@/functions";

const data: Array<grammar> = [
  {
    id: 1,
    name: "Страница 1",
    text: "Контент 1",
  },
  {
    id: 2,
    name: "Страница 2",
    text: "Контент 2",
  },
];

export async function grammarList() {
  await sleep(1000);
  return R.project(["id", "name"], data);
}

export const grammarDetail = async (id: number) => {
  await sleep(1000);
  return data.find((item) => item.id == id);
};

export async function updateDetail(id: number, updateData: grammar) {
  await sleep(1000);
  const update: grammar | undefined = data.find((item) => item.id == id);
  if (!update) return false;
  update.name = updateData.name;
  update.text = updateData.text;
  return true;
}
