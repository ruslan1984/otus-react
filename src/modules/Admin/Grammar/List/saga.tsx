import { takeEvery, put, call } from "redux-saga/effects";

import { grammarList } from "@grammar/data";
import { actions } from "./reducer";

export function* setGramarList() {
  yield put(actions.setLoading(true));
  const gl = yield call(grammarList);
  yield put(actions.fetch(gl));
  yield put(actions.setLoading(false));
}

export function* grammarListSaga() {
  yield takeEvery(actions.setList, setGramarList);
}
