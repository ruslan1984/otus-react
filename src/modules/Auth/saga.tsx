import { takeEvery, put, call, fork } from "redux-saga/effects";

import {
  login as sessionLogin,
  logout as sessionLogout,
  isAuthorised,
} from "./session";
import { CheckState } from "./types";
import { login as dataLogin } from "./data";
import { actions } from "./reducer";

export function* checkUserSession() {
  const auth: boolean = yield call(isAuthorised);
  if (auth) {
    yield put(actions.setState(CheckState.succeed));
  } else {
    yield put(actions.setState(CheckState.initiated));
  }
}

export function* auth(data: ReturnType<typeof actions.auth>) {
  try {
    yield put(actions.setState(CheckState.loading));
    const { user, password } = data.payload;
    if (user.trim() === "" || password.trim() === "") {
      return put(actions.setState(CheckState.failed));
    }
    const logined: boolean = yield call(dataLogin, user, password);
    if (logined) {
      yield call(sessionLogin, user);
      yield put(actions.login());
    } else {
      yield put(actions.setState(CheckState.failed));
    }
  } catch (error) {
    yield put(actions.logout());
  }
}
export function* logout() {
  yield call(sessionLogout);
  yield put(actions.logout());
}

export function* loginSaga() {
  yield fork(checkUserSession);
  yield takeEvery(actions.auth.type, auth);
  yield takeEvery(actions.logout.type, logout);
}
