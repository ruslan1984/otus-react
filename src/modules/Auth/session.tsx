import { sleep } from "@/functions";
import { data } from "./data";

export const login = async (user: string) => {
  await sleep(200);
  localStorage.setItem("auth", user);
};
export const logout = async () => {
  await sleep(200);
  localStorage.removeItem("auth");
};
export const isAuthorised = async () => {
  await sleep(200);
  return localStorage.getItem("auth") === data.user;
};
