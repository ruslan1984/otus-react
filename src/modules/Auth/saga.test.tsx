import { call } from "redux-saga/effects";
import { checkUserSession, auth, logout } from "./saga";
import { data, login as dataLogin } from "./data";
import { isAuthorised } from "./session";
import { CheckState } from "./types";

describe("Auth saga", (): void => {
  it("checkUserSession", async () => {
    const generator = checkUserSession();
    expect(generator.next().value).toEqual(call(isAuthorised));
    expect(generator.next().value).toMatchSnapshot();
    expect(generator.next().done).toBe(true);
  });
  it("login ok", async () => {
    const generator = auth({
      payload: { user: data.user, password: data.password },
    });

    expect(generator.next().value.payload.action).toEqual({
      type: "auth/setStatus",
      payload: CheckState.loading,
    });
    const logined = generator.next().value;
    expect(logined).toEqual(call(dataLogin, data.user, data.password));
    generator.next(logined);
    expect(generator.next().value.payload.action).toEqual({
      type: "auth/login",
      payload: undefined,
    });
    expect(generator.next().done).toBe(true);
  });
  it("login not ok", async () => {
    const generator = auth({
      payload: { user: data.user + "1", password: data.password },
    });
    expect(generator.next().value.payload.action).toEqual({
      type: "auth/setStatus",
      payload: CheckState.loading,
    });
    expect(generator.next().value).toEqual(
      call(dataLogin, data.user + "1", data.password)
    );
    generator.next(false);
    expect(generator.next().done).toBe(true);
  });
  it("logout", async () => {
    const generator = logout();
    expect(generator.next().value).toMatchSnapshot();
    expect(generator.next().value.payload.action).toEqual({
      type: "auth/logout",
      payload: undefined,
    });
    expect(generator.next().done).toBe(true);
  });
});
