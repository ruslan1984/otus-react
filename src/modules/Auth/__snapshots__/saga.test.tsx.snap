// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth saga checkUserSession 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 0,
      "type": "auth/setStatus",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`Auth saga logout 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
